### ISSUE TITLE
Ambiguity when referencing methods in module scope from within classes

### ISSUE THREAD

**Issue by foxbunny** [somebee/imba#154]
Consider this code:

```imba
def foo
    yes

class Foo
    def initialize
        foo
```

In the class `Foo` it is impossible to reference `foo` because it is always treated as `self.foo`.

Even if we use `var def`, we can run into the following ambiguity:

```imba
var def foo
    yes

class Foo
    def initialize
        foo  # which foo is this?

    def foo
       yes
```

I believe the cleanest solution to this would be to simply require that `self` is always used within a class, although that would probably be a solution for 2.0.

**Comment by somebee** [comment]
I don't agree :) Self should be implicit. It's not that ambiguous; if there is a variable of that name in scope (including outer scopes), it will always be a reference to that. If not, implicit self kicks in. Before 1.4 it was more ambiguous in the sense that def foo on root was treated like a weird semi-variable, so that when you referenced it inside a class it would always call that outer foo directly.

Better editor plugins is needed though. In sublime-imba we actually highlight variable references and calls in different colors, which makes it much easier to see.

**Comment by somebee** [comment]
The plugin is just helping the user see that there is an outer variable. Since Imba is not statically typed, we cannot for sure know whether there is a method of that name on class or not. And if it should never take precedence, how would you ever be able to refer to outer variables inside a class / method? :)

**Comment by somebee** [comment]
Variable declarations are not hoisted in Imba, so whenever you reference a variable before it is declared, the reference is treated as a method call. But you could be defining a subclass of some other class you have imported from somewhere else. Imba isn't able to analyse that on compile-time to know what methods are available in your class. I'm open for looking at explicit self, but as you said, it would have to be for a major new version later.
